GeneratorInterface/Core/interface/HadronizerFilter.h

```diff
@@ -269,6 +282,25 @@ namespace edm
       finalEvent->weights()[0] *= multihadweight;
     }
     
+    if (secondaryUnweighting_) {
+      double weight = finalGenEventInfo->weights()[0];
+      double absweight = std::abs(weight);
+      if (absweight<maxAbsWeight_) {
+        double randomflat = randomEngineSentry.randomEngine()->flat();
```

Comment:
What you have done looks good to me. The engine is associated with the streamID and module label. The only purpose of the sentry is to set the engine pointer to null when the filter method ends so the hadronizer cannot use it anymore. This prevents using the engine at some other time. Multiple sentries can point at the same engine and the top level filter function should also use the same engine. You could get the engine directly from the service using the streamID, but it would be the exactly same engine and take more lines of code and more CPU to find it. What you did is better.